### ISSUE TITLE
Keycloak SPI for Github org membership filtering

### ISSUE THREAD

**Issue by shaidar** [mitodl/ol-infrastructure#1651]
# Description/Context

When using Github as an Identity Provider in Keycloak, anyone with a Github account would be able to login to Keycloak. That is not the ideal situation for us and we would like to only allow members of the Github `mitodl` organization to login. It would also be beneficial to filter on team membership in a specific Github org.
# Acceptance Criteria

- [ ] Github IDP in Keycloak that only allows members of the `mitodl` organization to login and/or create an account in Keycloak.
- [ ] A clear error message to the user trying to login as to why they can't in case they don't belong to the right org and/or team

# Plan/Design

Based on Keycloak docs, the way to accomplish this is to create a custom [Service Provider Interface](https://www.keycloak.org/docs/latest/server_development/#_providers).

**Comment by shaidar** [comment]
I initially tried getting this done by writing a JS and using that script as part of the authentication flow following the steps [here](https://www.keycloak.org/docs/latest/server_development/#create-a-jar-with-the-scripts-to-deploy). After a good amount of testing with different error messages, it became clear that a script would not work. Apparently Keycloak uses Nashorn JavaScript Engine which is a limited environment and does not support making HTTP requests directly from scripts. So, the script is unable to make a Github API call to check a user's org membership or any other attribute such as team membership. The response from Github contains the following:
```
2023-07-26 14:12:16,889 DEBUG [org.keycloak.social.user_profile_dump] (executor-thread-8) User Profile JSON Data for provider github: {"login":"shaidar","id":xxxx,"node_id":"xxx","avatar_url":"https://avatars.githubusercontent.com/u/xxx?v=4","gravatar_id":"","url":"https://api.github.com/users/shaidar","html_url":"https://github.com/shaidar","followers_url":"https://api.github.com/users/shaidar/followers","following_url":"https://api.github.com/users/shaidar/following{/other_user}","gists_url":"https://api.github.com/users/shaidar/gists{/gist_id}","starred_url":"https://api.github.com/users/shaidar/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/shaidar/subscriptions","organizations_url":"https://api.github.com/users/shaidar/orgs","repos_url":"https://api.github.com/users/shaidar/repos","events_url":"https://api.github.com/users/shaidar/events{/privacy}","received_events_url":"https://api.github.com/users/shaidar/received_events","type":"User","site_admin":false,"name":"Sar","company":null,"blog":"","location":null,"email":null,"hireable":null,"bio":null,"twitter_username":null,"public_repos":8,"public_gists":0,"followers":3,"following":0,"created_at":"2012-02-17T17:22:54Z","updated_at":"2023-04-13T13:12:40Z"}
```

**Comment by shaidar** [comment]
I don't think this is useful anymore as ideally Keycloak should be the source of truths for accounts.

**Comment by blarghmatey** [comment]
I agree that unifying identity into Keycloak is the preferred approach.
